src/providers/azurestack/mod.rs

```diff
@@ -0,0 +1,369 @@
+
+        if versions.supported.versions.iter().any(|v| v == version) {
+            Ok(())
+        } else {
+            Err(anyhow!(
+                "fabric version '{}' not supported by the WireServer at '{}'",
+                version,
+                self.endpoint
+            ))
+        }
+    }
+
+    #[cfg(test)]
+    fn metadata_endpoint() -> String {
+        mockito::server_url()
+    }
+
+    #[cfg(not(test))]
+    fn metadata_endpoint() -> String {
+        const URL: &str = "http://169.254.169.254";
+        URL.to_string()
+    }
+
+    // Fetch the certificate.
+    fn fetch_cert(&self, certs_endpoint: String, mangled_pem: impl AsRef<str>) -> Result<String> {
+        let certs: goalstate::CertificatesFile = self
+            .client
+            .get(retry::Xml, certs_endpoint)
+            .header(
+                HeaderName::from_static(HDR_CIPHER_NAME),
+                HeaderValue::from_static("DES_EDE3_CBC"),
+            )
+            .header(
+                HeaderName::from_static(HDR_CERT),
+                HeaderValue::from_str(mangled_pem.as_ref())?,
+            )
+            .send()
+            .context("failed to get certificates")?
+            .ok_or_else(|| anyhow!("failed to get certificates: not found"))?;
+
+        // the cms decryption expects it to have MIME information on the top
+        // since cms is really for email attachments....
+        let mut smime = String::from(SMIME_HEADER);
+        smime.push_str(&certs.data);
+
+        Ok(smime)
+    }
+
+    // put it all together
+    fn get_ssh_pubkey(&self, certs_endpoint: String) -> Result<PublicKey> {
+        // we have to generate the rsa public/private keypair and the x509 cert
+        // that we use to make the request. this is equivalent to
+        // `openssl req -x509 -nodes -subj /CN=LinuxTransport -days 365 -newkey rsa:2048 -keyout private.pem -out cert.pem`
+        let (x509, pkey) = x509::generate_cert(&x509::Config::new(2048, 365))
+            .context("failed to generate keys")?;
+
+        // mangle the pem file for the request
+        let mangled_pem = crypto::mangle_pem(&x509).context("failed to mangle pem")?;
+
+        // fetch the encrypted cms blob from the certs endpoint
+        let smime = self
+            .fetch_cert(certs_endpoint, mangled_pem)
+            .context("failed to fetch certificate")?;
+
+        // decrypt the cms blob
+        let p12 = crypto::decrypt_cms(smime.as_bytes(), &pkey, &x509)
+            .context("failed to decrypt cms blob")?;
+
+        // convert that to the OpenSSH public key format
+        let ssh_pubkey = crypto::p12_to_ssh_pubkey(&p12)
+            .context("failed to convert pkcs12 blob to ssh pubkey")?;
+
+        Ok(ssh_pubkey)
+    }
+
+    fn fetch_hostname(&self) -> Result<Option<String>> {
+        let instance_metadata = AzureStack::fetch_identity();
+        Ok(Some(instance_metadata.unwrap().vm_name))
```

Comment:
The linter is indirectly saying to use `Err` instead of `unwrap()`, e.g.:

```suggestion
        let instance_metadata = AzureStack::fetch_identity()
        	.context("fetching instance identity")?;
        Ok(Some(instance_metadata.vm_name))
```

Reply:
Is there a way I should change `fetch_identity`: https://github.com/coreos/afterburn/pull/561/files#diff-34cbce01c9aa769653cb8f06a598911f28d90de5710e9b602560660669d702d1R172-R183 where I already have `context`?

Reply:
The `fetch_identity` function looks good. The `context` stuff is like a human-friendly backtrace where you can add more error-prefixing as the error bubbles up the stack to make it easier to tell where the error came from. It's up to you when to add a `.context(...)?` vs just `?`. If you think e.g. it's repetitive with the context already added by the callee, you can omit it in the caller. In this case for example, yeah it's probably superfluous given that my suggestion is more or less repeating the context provided by `fetch_identity`, and our caller ultimately adds the context that the failure happened while fetching the hostname:

```rust
        // write hostname if configured to do so
        self.hostname_file
            .map_or(Ok(()), |x| metadata.write_hostname(x))
            .context("writing hostname")?;
```

So you can probably just swap out `.context("fetching instance identity")?` with just `?` in my suggestion above.